extensions/python/PythonScriptEngine.h

```diff
@@ -0,0 +1,145 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "PythonBindings.h"
+#include "PyException.h"
+
+#include <mutex>
+#include <memory>
+#include <utility>
+#include <exception>
+#include <string>
+#include <vector>
+
+#include "core/ProcessSession.h"
+#include "core/Processor.h"
+
+#include "PythonScriptProcessContext.h"
+#include "PythonProcessor.h"
+#include "types/PyProcessSession.h"
+#include "PythonScriptException.h"
+
+#if defined(__GNUC__) || defined(__GNUG__)
+#pragma GCC visibility push(hidden)
+#endif
+
+namespace org::apache::nifi::minifi::extensions::python {
+
+#if defined(__GNUC__) || defined(__GNUG__)
+class __attribute__((visibility("default"))) GlobalInterpreterLock {
+#else
+class GlobalInterpreterLock {
+#endif
+ public:
+  GlobalInterpreterLock();
+  ~GlobalInterpreterLock();
+
+ private:
+  PyGILState_STATE gil_state_;
+};
+
+struct Interpreter {
+  Interpreter();
```

Comment:
can multiple `Interpreter` instances coexit? if not we could make the constructor to be only callable from `getInterpreter`

Reply:
You are right, I've made the ctor private and changed the getInterpreter to be a static function of the Interpreter class in https://github.com/apache/nifi-minifi-cpp/pull/1504/commits/13a0c350eaaa7bd886d398204fd19fb2abf2870d